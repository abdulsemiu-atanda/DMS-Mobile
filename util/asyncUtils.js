import sendRequest from '../requests/requestUtil'

/**
 * Constructs async actions constants
 * @param {string} baseName
 * @returns {Object}
 */
export const asyncActionNames = baseName => ({
  failure: `${baseName}_FAILURE`,
  loading: `${baseName}_LOADING`,
  success: `${baseName}_SUCCESS`
})

/**
 * Constructs Redux async actions
 * @param {string} actionName
 * @returns {Object}
 */
export const asyncActions = (actionName) => ({
  loading: bool => ({
    type: asyncActionNames(actionName).loading,
    data: bool
  }),
  failure: (bool, error) => ({
    type: asyncActionNames(actionName).failure,
    data: {error, status: bool}
  }),
  success: data => ({
    type: asyncActionNames(actionName).success,
    data
  })
})

/**
 * Dispatches Redux Actions
 * @async asyncRequest
 * @param {string} ACTION_NAME - The name of action to be dispatched
 * @param {string} endpoint - The endpoint to resource
 * @param {string} method - HTTP method
 * @param {Object} data - HTTP request body
 * @param {string} token - Authentication token for request
 * @returns {Function} - Function that dispatches Redux Actions
 */

export const asyncRequest = (ACTION_NAME, endpoint, method, data, token) => dispatch => {
  dispatch(asyncActions(ACTION_NAME).loading(true))
  sendRequest(endpoint, method, data, token)
    .then(res => res.json())
    .then(res => {
      dispatch(asyncActions(ACTION_NAME).success(res))
      dispatch(asyncActions(ACTION_NAME).loading(false))
    })
    .catch(err => dispatch(asyncActions(ACTION_NAME).failure(true, err)))
}
