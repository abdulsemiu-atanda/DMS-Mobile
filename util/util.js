import {
  emailCorrect,
  emailError,
  fullnameCorrect,
  fullnameError,
  passwordMatch,
  passwordMatchError,
  usernameCorrect,
  usernameError
} from '../actions/formActions'

export const validateFields = data => {
  return dispatch => {
    Object.keys(data).forEach(field => {
      switch (field) {
        case 'email':
          return /^\S+@\S+$/g.test(data[field]) ? dispatch(emailCorrect(true)) : dispatch(emailError(true))
        case 'username':
          return data[field].trim() === '' ? dispatch(usernameError(true)) : dispatch(usernameCorrect(true))
        case 'password':
          return data.password === data.confirmPassword ? dispatch(passwordMatch(true)) : dispatch(passwordMatchError(true))
        case 'fullname':
          return data[field].trim() === '' ? dispatch(fullnameError(true)) : dispatch(fullnameCorrect(true))
        default:
          break
      }
    })
  }
}