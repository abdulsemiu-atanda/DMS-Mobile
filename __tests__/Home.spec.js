import 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'
import {fromJS} from 'immutable'

import {Home} from '../components/Home.react'
import {noop} from '../util/util'

const publicDocuments = [
  {
    title: 'Scooby Doo',
    access: 'public',
    content: 'Scooby Doo is a beloved brown dog.',
    updatedAt: Date.now(),
    createdAt: Date.now()
  },
  {
    title: 'Sharpie',
    access: 'public',
    content: 'Sharpie is one of the perfectly crafted products.',
    updatedAt: Date.now(),
    createdAt: Date.now()
  }
]

const props = {
  asyncRequest: noop,
  document: {
    documentLoading: false,
    documents: fromJS(publicDocuments)
  },
  navigation: {
    state: {
      routeName: 'All'
    }
  },
  user: {
    documents: fromJS(publicDocuments),
    loadingDocuments: false
  }
}

describe('Home', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<Home {...props} />).toJSON()

    expect(tree).toMatchSnapshot()
    expect(tree.children.length).toEqual(2)
    expect(tree.children[0].children[0].children[0].children.length).toEqual(publicDocuments.length)
  })

  it('renders correctly when user is viewing their documents', () => {
    const userViewProps = {
      ...props,
      navigation: {
        state: {
          routeName: 'Collections'
        }
      }
    }
    const tree = renderer.create(<Home {...userViewProps}  />).toJSON()

    expect(tree).toMatchSnapshot()
    expect(tree.children.length).toEqual(2)
  })

  it('does not render add button when there are no documents', () => {
    const noDocProp = {
      ...props,
      document: {
        documentLoading: false,
        documents: fromJS([])
      }
    }
    const tree = renderer.create(<Home {...noDocProp} />).toJSON()

    expect(tree).toMatchSnapshot()
    expect(tree.children.length).toEqual(3)
  })

  it('renders spinner when document is loading', () => {
    const loadingDocProp = {
      ...props,
      document: {
        documentLoading: true,
        documents: fromJS([])
      }
    }
    const tree = renderer.create(<Home {...loadingDocProp} />).toJSON()

    expect(tree).toMatchSnapshot()
    expect(tree.type).toEqual('ActivityIndicator')
  })
})

