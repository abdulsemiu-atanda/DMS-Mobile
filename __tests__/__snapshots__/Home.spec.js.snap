// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home does not render add button when there are no documents 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={false}
    ellipsizeMode="tail"
    style={
      Array [
        Object {
          "color": "grey",
          "fontSize": 70,
        },
        undefined,
        Object {
          "fontFamily": "Ionicons",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
      ]
    }
  >
    
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontFamily": "Verdana",
        "marginBottom": 20,
        "textAlign": "center",
      }
    }
  >
    Oops! there are no public documents currently available for viewing.        Please check again later or start by adding a public document.
  </Text>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignContent": "center",
        "alignSelf": "center",
        "backgroundColor": "#0983ff",
        "padding": 7,
        "width": 300,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#aacfff",
          "fontSize": 15,
          "textAlign": "center",
        }
      }
    >
      ADD DOCUMENT
    </Text>
  </View>
</View>
`;

exports[`Home renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#f7f7f7",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "padding": 7,
      }
    }
  >
    <RCTScrollView
      dataSource={
        ListViewDataSource {
          "items": 2,
        }
      }
      enableEmptySections={true}
      renderRow={[Function]}
      renderScrollComponent={[Function]}
      style={
        Object {
          "backgroundColor": "#f7f7f7",
        }
      }
    >
      <View>
        <View
          style={
            Object {
              "borderRadius": 5,
              "marginBottom": 5,
              "width": 750,
            }
          }
        >
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Object {
                  "backgroundColor": "#fff",
                  "height": 266.8,
                  "padding": 10,
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Object {
                    "fontSize": 15,
                    "fontWeight": "700",
                    "marginBottom": 2,
                  }
                }
              >
                Scooby Doo
              </Text>
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                numberOfLines={3}
                style={
                  Object {
                    "marginBottom": 106.72,
                  }
                }
              >
                Scooby Doo is a beloved brown dog.
              </Text>
              <View
                style={
                  Object {
                    "borderBottomColor": "#bbb",
                    "borderBottomWidth": 1,
                    "marginBottom": 2,
                  }
                }
              />
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Object {
                    "color": "#bbb",
                    "fontSize": 10,
                  }
                }
              >
                Modified: 
                Jun 4th 18
                 / Created: 
                Jun 4th 18
              </Text>
            </View>
          </View>
        </View>
        <View
          style={
            Object {
              "borderRadius": 5,
              "marginBottom": 5,
              "width": 750,
            }
          }
        >
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Object {
                  "backgroundColor": "#fff",
                  "height": 266.8,
                  "padding": 10,
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Object {
                    "fontSize": 15,
                    "fontWeight": "700",
                    "marginBottom": 2,
                  }
                }
              >
                Sharpie
              </Text>
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                numberOfLines={3}
                style={
                  Object {
                    "marginBottom": 106.72,
                  }
                }
              >
                Sharpie is one of the perfectly crafted products.
              </Text>
              <View
                style={
                  Object {
                    "borderBottomColor": "#bbb",
                    "borderBottomWidth": 1,
                    "marginBottom": 2,
                  }
                }
              />
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Object {
                    "color": "#bbb",
                    "fontSize": 10,
                  }
                }
              >
                Modified: 
                Jun 4th 18
                 / Created: 
                Jun 4th 18
              </Text>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "#0983ff",
        "borderRadius": 33.75,
        "height": 67.5,
        "justifyContent": "center",
        "left": 660,
        "position": "absolute",
        "top": 1067.2,
        "width": 67.5,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={false}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#aacfff",
            "fontSize": 30,
          },
          Object {
            "alignSelf": "center",
          },
          Object {
            "fontFamily": "Ionicons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
        ]
      }
    >
      
    </Text>
  </View>
</View>
`;

exports[`Home renders correctly when user is viewing their documents 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#f7f7f7",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "padding": 7,
      }
    }
  >
    <RCTScrollView
      dataSource={
        ListViewDataSource {
          "items": 2,
        }
      }
      enableEmptySections={true}
      renderRow={[Function]}
      renderScrollComponent={[Function]}
      style={
        Object {
          "backgroundColor": "#f7f7f7",
        }
      }
    >
      <View>
        <View
          style={
            Object {
              "borderRadius": 5,
              "marginBottom": 5,
              "width": 750,
            }
          }
        >
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Object {
                  "backgroundColor": "#fff",
                  "height": 266.8,
                  "padding": 10,
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Object {
                    "fontSize": 15,
                    "fontWeight": "700",
                    "marginBottom": 2,
                  }
                }
              >
                Scooby Doo
              </Text>
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                numberOfLines={3}
                style={
                  Object {
                    "marginBottom": 106.72,
                  }
                }
              >
                Scooby Doo is a beloved brown dog.
              </Text>
              <View
                style={
                  Object {
                    "borderBottomColor": "#bbb",
                    "borderBottomWidth": 1,
                    "marginBottom": 2,
                  }
                }
              />
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Object {
                    "color": "#bbb",
                    "fontSize": 10,
                  }
                }
              >
                Modified: 
                Jun 4th 18
                 / Created: 
                Jun 4th 18
              </Text>
            </View>
          </View>
        </View>
        <View
          style={
            Object {
              "borderRadius": 5,
              "marginBottom": 5,
              "width": 750,
            }
          }
        >
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Object {
                  "backgroundColor": "#fff",
                  "height": 266.8,
                  "padding": 10,
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Object {
                    "fontSize": 15,
                    "fontWeight": "700",
                    "marginBottom": 2,
                  }
                }
              >
                Sharpie
              </Text>
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                numberOfLines={3}
                style={
                  Object {
                    "marginBottom": 106.72,
                  }
                }
              >
                Sharpie is one of the perfectly crafted products.
              </Text>
              <View
                style={
                  Object {
                    "borderBottomColor": "#bbb",
                    "borderBottomWidth": 1,
                    "marginBottom": 2,
                  }
                }
              />
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Object {
                    "color": "#bbb",
                    "fontSize": 10,
                  }
                }
              >
                Modified: 
                Jun 4th 18
                 / Created: 
                Jun 4th 18
              </Text>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "#0983ff",
        "borderRadius": 33.75,
        "height": 67.5,
        "justifyContent": "center",
        "left": 660,
        "position": "absolute",
        "top": 1067.2,
        "width": 67.5,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={false}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#aacfff",
            "fontSize": 30,
          },
          Object {
            "alignSelf": "center",
          },
          Object {
            "fontFamily": "Ionicons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
        ]
      }
    >
      
    </Text>
  </View>
</View>
`;

exports[`Home renders spinner when document is loading 1`] = `
<ActivityIndicator
  animating={true}
  color="blue"
  hidesWhenStopped={true}
  size="large"
/>
`;
